src/DataTypes/DataTypesNumber.h

```diff
@@ -9,10 +10,21 @@
 namespace DB
 {
 
+using DataTypes = std::vector<DataTypePtr>;
+
 template <typename T>
 class DataTypeNumber final : public DataTypeNumberBase<T>
 {
 public:
+    DataTypeNumber() = default;
+
+    explicit DataTypeNumber(DataTypePtr opposite_sign_data_type_)
```

Comment:
Actually, I am not sure it's good to modify general `DataTypeNumber`/`IDataType` class for using it only in one specific place. It just could be confusing why it's needed for future readers. I think it will be better to add a separate class derived from `DataTypeNumber` for this perpose. We can name it like `DataTypeBothSignsNumber` and add a comment why we need it and where it's used (I am not good at names, you can think of better one). And in `getLeastSupertype` instead of calling `hasOppositeSignDataType` we will just try to cast type to `DataTypeBothSignsNumber` (using `typeid_cast`/`dynamic_cast` for example, no need for new methods in `IDataType`).